lib/Dialect/FIRRTL/Transforms/ExpandWhens.cpp

```diff
@@ -103,18 +111,134 @@ class LastConnectResolver : public FIRRTLVisitor<ConcreteT> {
   using FIRRTLVisitor<ConcreteT>::visitDecl;
   using FIRRTLVisitor<ConcreteT>::visitStmt;
 
+  // This function simplifies bitindex expressions by pushing them through cat
+  // and mux ops. For example `cat(a, b)[1:0]` where b's width is >= 2 is just
+  // b[1:0]. And `mux(sel, a, b)[3:0]` is `mux(sel, a[3:0], b[3:0])`. This
+  // canonicalization is necessary to eliminate uses of subword uninitialized
+  // wire ops so that initialization checking only signals a failure if a bit
+  // is truly not assigned.
+  Value canonicalizeBits(BitsPrimOp op, ImplicitLocOpBuilder builder) {
+    if (!op.getInput().getDefiningOp()) {
+      return op;
+    }
+    auto *inputOp = op.getInput().getDefiningOp();
+    if (auto cat = dyn_cast<CatPrimOp>(inputOp)) {
+      // bits(cat(a, b), ...) => bits(a, ...), or bits(b, ...), or cat(bits(a,
+      // ...), bits(b, ...))
+      if (auto rhsT = cat.getRhs().getType().dyn_cast<IntType>()) {
+        uint32_t lhsLo = rhsT.getWidth().getValue();
+        uint32_t rhsHi = lhsLo - 1;
+        Value newOp;
+        if (op.getHi() >= lhsLo && op.getLo() >= lhsLo) {
+          // only indexing the lhs
+          newOp = canonicalizeBits(
+              builder.create<BitsPrimOp>(cat.getLhs(), op.getHi() - lhsLo,
+                                         op.getLo() - lhsLo),
+              builder);
+        } else if (op.getHi() <= rhsHi && op.getLo() <= rhsHi) {
+          // only indexing the rhs
+          newOp = canonicalizeBits(
+              builder.create<BitsPrimOp>(cat.getRhs(), op.getHi(), op.getLo()),
+              builder);
+        } else {
+          auto bitsLhs = canonicalizeBits(
+              builder.create<BitsPrimOp>(cat.getLhs(), op.getHi() - lhsLo, 0),
+              builder);
+          auto bitsRhs = canonicalizeBits(
+              builder.create<BitsPrimOp>(cat.getRhs(), rhsHi, op.getLo()),
+              builder);
+          newOp = builder.createOrFold<CatPrimOp>(bitsLhs, bitsRhs);
+        }
+        return newOp;
+      }
+    } else if (auto mux = dyn_cast<MuxPrimOp>(op.getInput().getDefiningOp())) {
+      // bits(mux(sel, a, b), ...) => mux(sel, bits(a, ...), bits(b, ...))
+      auto bitsHigh = canonicalizeBits(
+          builder.create<BitsPrimOp>(mux.getHigh(), op.getHi(), op.getLo()),
+          builder);
+      auto bitsLow = canonicalizeBits(
+          builder.create<BitsPrimOp>(mux.getLow(), op.getHi(), op.getLo()),
+          builder);
+      Value newOp =
+          builder.createOrFold<MuxPrimOp>(mux.getSel(), bitsHigh, bitsLow);
+      return newOp;
+    }
+    return op;
+  }
+
+  // Rewrites a connection where the dest is the bitindex. An expression that
+  // looks like `x[hi:lo] <= e` gets rewritten to `x <= cat(xprev[x.w-1:hi+1],
+  // e, xprev[lo-1:0])`, where xprev is the most recent connection to x in any
+  // scope.
+  void rewriteBitsConnect(BitsPrimOp bits, FieldRef &dest,
+                          FConnectLike &connection) {
+    // get x's width
```

Comment:
Sorry for the nitpicking but we usually start our comments with a capital, and end in a period.